### PR TITLE
Add information about removable settings to the deprecation info API

### LINKED ISSUE #83116: [Deprecation Info API] Provide enough meta data to allow Kibana Upgrade Assistant remove dynamic settings.
The Deprecation Info API from Elasticsearch provides much of the source information for Kibana Upgrade assistant. For a very small subset of deprecated dynamic settings Kibana Upgrade Assistant can unset/remove these from the Elasticsearch cluster with a simple button click. A small subset is only supported today due to a lack of meta data needed to generalize the removal of these type of settings. In https://github.com/elastic/elasticsearch/issues/73089 the API gained support for generic meta data. 

This issue is to define the meta data needed to allow Kibana Upgrade Assistant to generalize the ability to remove/unset dynamic settings with a single button click. 

Since meta data is arbitrary key/value pairing... I would suggest something like

```json
...
  "_meta": {
    "actions" : {
      "remove_settings" : {
        "settings" : ["foo", "bar"]
    }
  }
...
```
This structure allows to add in additional future actions, and provides the context for that action. Since a single deprecation info message may have multiple settings that are deprecated, `settings` must be an array. In this case "foo" and "bar" would be the fully qualified setting name (e.g. `translog.retention.size`) and would be guaranteed to be able to be set/unset dynamically. Each deprecation message would have a remove_settings action iff the deprecation warning can be fully resolved by un-setting those settings. Some settings can be set both statically (elasticsearch.yml) and dynamically (via the API), the settings listed here would be guaranteed to have been set only dynamically (and thus can be unset dynamically). If a setting was actually set both statically and dynamically the deprecation message would not have the "remove_settings" entry since dynamically un-setting the setting would not resolve the deprecation warning. Further, there might be edge cases w.r.t. dependent settings or the scope of a deprecation warning..in all cases Elasticsearch must guarantee that no additional context/information is needed and un-setting the these setting listed dynamically will fully resolve the entire deprecation warning. 

When Kibana receives a deprecation message with the remove_settings action, it can be assured that taking that single action will resolve the full deprecation message. To remove/un-set the settings, Kibana will need to set the value to `null` for both persistent and transient settings. 

Alternatively, "actions" (or the like) could be a new top level item, reserving _meta for less structured information. However, I have a minor preference to keep it under _meta .

cc @alisonelizabeth @masseyke 

related: https://github.com/elastic/kibana/pull/123599

### PR DESCRIPTION
**Pull request by masseyke**
Deprecated settings that are only set dynamically (i.e. not in elasticsearch.yml) could also be removed dynamically. This
PR adds information about these settings to the deprecation info API.
Closes #83116

### CODE REVIEW

**Comment by masseyke** [comment]
Converted to a draft because I realized a shortcoming -- if a property is in elasticsearch.yml on one node but not on another (and it's also set as a dynamic setting), you'll get two separate DeprecationIssues in the response -- one saying you can fix it dynamically and one not saying that.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/deprecation/DeprecationIssue.java [hunk 1]
```diff
@@ -175,4 +185,103 @@ public int hashCode() {
...
+
+    /**
+     * This method returns a DeprecationIssue that has in its meta object the intersection of all auto-removable settings that appear on
+     * all of the DeprecationIssues that are passed in. This method assumes that all DeprecationIssues passed in are equal, except for the
+     * auto-removable settings in the meta object.
+     * @param similarIssues
+     * @return
+     */
+    @SuppressWarnings("unchecked")
+    public static DeprecationIssue getIntersectionOfRemovableSettings(List<DeprecationIssue> similarIssues) {
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/deprecation/DeprecationIssue.java:204]: This function feels really large, could you split it up into smaller pieces? Maybe a separate part for retrieving the leastCommonRemovableSettings?

**Review by dakrone — CHANGES_REQUESTED** [review]
Thanks for working on this Keith, I tried looking at this multiple times, but each time I found it hard to follow the entire logic through (not your fault, just because of the generic maps all over).

What do you think of my suggestion to make the actions and settings removal action first-class objects? I think it might make it easier to read through the flow of the code.

**Comment by masseyke** [comment]
@dakrone I've broken up all long methods and tried to document what's going on a little better. Let me know if it makes more sense now.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/deprecation/DeprecationIssue.java (hunk 1 above)
Comment (masseyke) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/deprecation/DeprecationIssue.java:201]: Done.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/deprecation/DeprecationIssue.java [hunk 2]
```diff
@@ -175,4 +190,233 @@ public int hashCode() {
...
+     * @return A DeprecationIssue containing only the removal actions that are in all similarIssues
+     */
+    public static DeprecationIssue getIntersectionOfRemovableSettings(List<DeprecationIssue> similarIssues) {
+        if (similarIssues == null || similarIssues.isEmpty()) {
+            return null;
+        }
+        if (similarIssues.size() == 1) {
+            return similarIssues.get(0);
+        }
+        DeprecationIssue representativeIssue = similarIssues.stream().findAny().get();
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/deprecation/DeprecationIssue.java:212]: Seems like this could be `similarIssues.get(0)` since it's not filtering or anything?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/deprecation/DeprecationIssue.java [hunk 3]
```diff
@@ -175,4 +190,233 @@ public int hashCode() {
...
+            }
+            return new Meta(actionsIntersection, nonActionMetadata);
+        }
+
+        /*
+         * Returns an Optional Meta object from a DeprecationIssue's meta Map. If the meta Map is null then the Optional will not be
+         * present.
+         */
+        @SuppressWarnings("unchecked")
+        public static Optional<Meta> fromMetaMap(Map<String, Object> metaMap) {
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/deprecation/DeprecationIssue.java:323]: I think this can be `private` since it's only used from this same file

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/deprecation/DeprecationIssue.java [hunk 4]
```diff
@@ -175,4 +190,233 @@ public int hashCode() {
...
+            if (actionMaps == null) {
+                actions = null;
+            } else {
+                actions = new ArrayList<>();
+                for (Map<String, Object> actionMap : actionMaps) {
+                    final Action action;
+                    if (REMOVE_SETTINGS_ACTION_TYPE.equals(actionMap.get(ACTION_TYPE))) {
+                        action = RemovalAction.fromActionMap(actionMap);
+                    } else {
+                        action = UnknownAction.fromActionMap(actionMap);
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/deprecation/DeprecationIssue.java:338]: Do we need the `UnknownAction` at all, or can we be strict and throw an exception if it's not in the format we expect? Do you think that would be too strict here? (especially since this `_meta` map the user has no control over, so I think we'd only be limiting plugins)

**Review by dakrone — APPROVED** [review]
I think this LGTM, I left three pretty minor comments.

I do think we incur a bit of tech debt for this, but since it's 7.17 I think it's reasonable (tell me if you disagree!). I think that we will need to plan for this with whatever code we want to do this behavior with in the 8.x line.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/deprecation/DeprecationIssue.java (hunk 2 above)
Comment (masseyke) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/deprecation/DeprecationIssue.java:212]: Yeah that's what I originally had it as, but I changed it to make it clear that it could be any (the first one's not special). I can put it back. I flip-flopped a couple of times on this and am not tied to either way.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/deprecation/DeprecationIssue.java (hunk 4 above)
Comment (masseyke) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/deprecation/DeprecationIssue.java:338]: Yeah I was afraid I'd break some plugin that I don't know about. If you're certain that no one else is putting in any other `action`s, I can remove this.

**Comment by masseyke** [comment]
> I think this LGTM, I left three pretty minor comments.
> 
> I do think we incur a bit of tech debt for this, but since it's 7.17 I think it's reasonable (tell me if you disagree!). I think that we will need to plan for this with whatever code we want to do this behavior with in the 8.x line.

Oh there's definitely new tech debt here due to the fact that `meta` is just a map of string to object, and that we're having to layer in meaning. I agree that it's fine for now, and better than the alternative of a major overhaul in 7.x. As we've discussed, we need to change `meta` into something more concrete in 8.x.
